crossbeam-skiplist/src/base.rs

```diff
@@ -1675,25 +1687,25 @@ where
 }
 
 /// An iterator over a subset of entries of a `SkipList`.
-pub struct Range<'a: 'g, 'g, 'k, Min, Max, K: 'a, V: 'a>
+pub struct Range<'a: 'g, 'g, T, R, K: 'a, V: 'a>
 where
-    K: Ord + Borrow<Min> + Borrow<Max>,
-    Min: Ord + ?Sized + 'k,
-    Max: Ord + ?Sized + 'k,
+    K: Ord + Borrow<T>,
+    R: RangeBounds<T>,
+    T: Ord + ?Sized,
 {
     parent: &'a SkipList<K, V>,
-    lower_bound: Bound<&'k Min>,
-    upper_bound: Bound<&'k Max>,
     head: Option<&'g Node<K, V>>,
     tail: Option<&'g Node<K, V>>,
+    range: R,
     guard: &'g Guard,
+    _marker: PhantomData<T>,
```

Comment:
I think it is needed so that the compiler knows what variance to put on `Q`. I think `PhantomData<*const Q>` might be more accurate in this case since we don't actually own a `Q`.

Reply:
That explains it.

Perhaps `PhantomData<fn(&R) -> Bound<&Q>>` would be even more accurate? :) 
The reason is that we have a `range: R` and only get a `Q` as `Bound<&Q>` returned from `range.start_bound()` and `range.end_bound()`.

Reply:
No, `*const Q` is correct in this case since we want to be covariant over `T`: this means that a `Range<T = &'a u32>` can be implicitly converted to a `Range<T = &'b u32>` where `b` is shorter than `a`.

Reply:
So if we have `PhantomData<fn(A) -> B>` we'll be contravariant over `A` and covariant over `B`.

I've simplified things a bit to `PhantomData<fn() -> Q>` so that should give us the intended variance?

This is a little easier than `*const T` because then we don't have to manually implement `Send` and `Sync` for `RefRange` due to raw pointers being `!Send` and `!Sync`.

Reply:
No, `PhantomData<fn() -> Q>` is wrong here. We want the bound to be covariant, not contravariant.

Reply:
If you don't want to deal with `Send`/`Sync` then just use `PhantomData<&'static Q>`.